### ISSUE TITLE
[C++][IWYU] How to treat includes in case of compiler deduction?

### ISSUE THREAD

**Issue by marcogmaia** [google/styleguide#863]
I have a question about header inclusion best practices when dealing with implicit type construction.

Consider the following code structure:

```cpp 
// a.h 
struct A { int x, y; };

// b.h 
#include "a.h"

struct B { void Func(A a) { /* use a */ } };

// c.cpp 
#include "b.h" 

void Func() { 
  B b; 
  b.Func({1, 2}); // Implicit construction of A 
}
```
In `c.cpp`, we're using struct `A` implicitly through brace initialization, and the compiler successfully deduces the type for us because `b.h` includes `a.h`.  This pattern is common when working with aggregate types.

### Questions:
- Should we explicitly include `a.h` in `c.cpp` even though we're not directly referencing the symbol?
- What's the general consensus for header inclusion when dealing with implicit type construction?

I'd appreciate insights on this scenario.

**Comment by parth-patel010** [comment]
Yes, you should explicitly include a.h in c.cpp if you rely on A (even implicitly).

1. Include What You Use Principle
This principle says:

Any file that uses a symbol must include the declaration of that symbol.

Even though you're not writing A by name, the code:

cpp
Copy
Edit
b.Func({1, 2});
constructs an A, which is a usage — just an implicit one.

If b.h ever stops including a.h, your code will break silently. That’s what “include what you use” protects against.

2. Improves Compilation Resilience
Your code is now fragile. It relies on a transitive include, which is:

Not guaranteed

Subject to internal changes in b.h

A potential source of subtle bugs or compile errors in future refactors

Explicit includes make your translation unit (c.cpp) resilient to unrelated changes in other files.

3. Makes Dependencies Clear
Imagine someone new reading c.cpp. The use of {1, 2} is obscure without knowing it's constructing an A. If a.h is included explicitly, it becomes clear that A is relevant to this file.

General Consensus in Modern C++
Google C++ Style Guide, LLVM, and Mozilla all encourage:

Directly include all headers that define symbols used in your file, even if they are also included indirectly.

IWYU (Include What You Use) tool would recommend including a.h in c.cpp because {1, 2} implies use of A.

 Final Verdict
 YES — include a.h in c.cpp
Even if the compiler doesn’t complain today, your code will be:

More stable

More maintainable

More understandable
